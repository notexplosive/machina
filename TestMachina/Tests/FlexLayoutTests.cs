using ApprovalTests;
using ApprovalTests.Reporters;
using Machina.Data;
using Machina.Data.Layout;
using Microsoft.Xna.Framework;
using TestMachina.Utility;
using Xunit;

namespace TestMachina.Tests
{
    [UseReporter(typeof(DiffReporter))]
    public class FlexLayoutTests
    {
        [Fact]
        public void can_create_flex_layout()
        {
            var layout = FlexLayout.HorizontalFlexParent("root", FlexLayoutStyle.Empty,
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 3)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(8, 5)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 8))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void can_create_vertical_flex_layout()
        {
            var layout = FlexLayout.VerticalFlexParent("root", FlexLayoutStyle.Empty,
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 3)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(8, 5)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 8))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flex_layout_supports_margin()
        {
            var layout = FlexLayout.HorizontalFlexParent("root", new FlexLayoutStyle(style: new LayoutStyle(margin: new Point(5, 5))),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 3)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(8, 5)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 8))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flex_layout_supports_padding()
        {
            var layout = FlexLayout.HorizontalFlexParent("root", new FlexLayoutStyle(style: new LayoutStyle(margin: new Point(5, 5), padding: 3)),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 3)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(8, 5)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 8))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flex_considers_alignment_when_things_have_varied_perpendicular_size()
        {
            var layout = FlexLayout.HorizontalFlexParent("root", new FlexLayoutStyle(style: new LayoutStyle(margin: new Point(5, 5), padding: 3, alignment: Alignment.BottomCenter)),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 3)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(8, 5)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 8))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flex_can_specify_a_min_width()
        {
            var layout = FlexLayout.HorizontalFlexParent("root", new FlexLayoutStyle(minAlongSize: 100, style: new LayoutStyle(margin: new Point(5, 5), padding: 3, alignment: Alignment.BottomCenter)),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 3)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(8, 5)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 8))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        [Fact]
        public void flex_can_specify_a_min_height()
        {
            var layout = FlexLayout.HorizontalFlexParent("root", new FlexLayoutStyle(minPerpendicularSize: 25, style: new LayoutStyle(margin: new Point(5, 5), padding: 3, alignment: Alignment.BottomCenter)),
                LayoutNode.Leaf("itemA", LayoutSize.Pixels(10, 3)),
                LayoutNode.Leaf("itemB", LayoutSize.Pixels(12, 10)),
                LayoutNode.Leaf("itemC", LayoutSize.Pixels(8, 5)),
                LayoutNode.Leaf("itemD", LayoutSize.Pixels(9, 10)),
                LayoutNode.Leaf("itemE", LayoutSize.Pixels(7, 8))
            );

            var result = layout.Bake();
            Approvals.Verify(LayoutNodeUtils.DrawResult(result));
        }

        // TODO: Max Width
        // TODO: Max Height
    }
}