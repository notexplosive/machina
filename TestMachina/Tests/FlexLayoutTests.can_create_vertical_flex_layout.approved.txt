111111111100
1itemA   1 0
1111111111 0
111111111111
1itemB     1
1          1
1          1
1          1
1          1
1          1
1          1
1          1
111111111111
11111111   0
1itemC 1   0
1      1   0
1      1   0
11111111   0
111111111  0
1itemD  1  0
1       1  0
1       1  0
1       1  0
1       1  0
1       1  0
1       1  0
1       1  0
111111111  0
1111111    0
1itemE1    0
1     1    0
1     1    0
1     1    0
1     1    0
1     1    0
111111100000
